### ISSUE TITLE
Improve the error message when adding documents

### ISSUE THREAD

**Issue by irevoire** [meilisearch/meilisearch#2680]
Currently, when adding json documents to meilisearch, you can either add them as a single document;
```bash
curl -X POST 'http://localhost:7700/indexes/mieli/documents'  -H 'Content-type: application/json' -d '{ "id": 1, "pet": "dog" }'
```
Or as an array of documents;
```bash
curl -X POST 'http://localhost:7700/indexes/mieli/documents'  -H 'Content-type: application/json' -d '[{ "id": 1, "pet": "dog" }, { "id": 2, "pet": "doggo" }]'
```

That's cool, but the error messages are not explicit at all;
```bash
curl -X POST 'http://localhost:7700/indexes/mieli/documents'  -H 'Content-type: application/json' -d '"sending a string instead of a json"'
```
For example, with this request we get this error message;
```
The `json` payload provided is malformed. `Couldn't serialize document value: data did not match any variant of untagged enum Either`.
```

I would like to improve this message and get rids entirely of the notion of « untagget enum » and « Either ».

**Comment by onyxcherry** [comment]
Have you encountered any other meaningless messages?

That issue doesn't seem to will be resolved by the serde team (as they have a low priority on it).

Although the message could be improved by implementing own `Deserialize` for `ArrayOrSingleObject`, isn't it better to simply replace the message:

```rust
let mut serde_msg = se.to_string();
if serde_msg == "data did not match any variant of untagged enum Either" {
    serde_msg = "data are neither an object nor a list of objects".to_string();
}
```
?

**Comment by inkss** [comment]
I encountered the same file upload error, how can I fix it?

```sh
curl -X POST 'https://xxxx/indexes/xxxx/documents/?primaryKey=xxxx' \
  -H "Authorization: Bearer ${{secrets.MEILISEARCH_KEY}}" \
  -H "Content-Type: application/json" \
  --data-binary 'content.json'
```

``` json
{
    "message": "The `json` payload provided is malformed. `Couldn't serialize document value: expected value at line 1 column 1`.",
    "code": "malformed_payload",
    "type": "invalid_request",
    "link": "https://docs.meilisearch.com/errors#malformed_payload"
}
```

**Comment by onyxcherry** [comment]
> I encountered the same file upload error, how can I fix it?
> 
> ```sh
> curl -X POST 'https://xxxx/indexes/xxxx/documents/?primaryKey=xxxx' \
>   -H "Authorization: Bearer ${{secrets.MEILISEARCH_KEY}}" \
>   -H "Content-Type: application/json" \
>   --data-binary 'content.json'
> ```
> 
> ``` json
> {
>     "message": "The `json` payload provided is malformed. `Couldn't serialize document value: expected value at line 1 column 1`.",
>     "code": "malformed_payload",
>     "type": "invalid_request",
>     "link": "https://docs.meilisearch.com/errors#malformed_payload"
> }
> ```

Add `@` sign before the file name.
So use `--data-binary "@path/to/file"` flag

**Comment by inkss** [comment]
Thank you!!!!! 😂

**Comment by irevoire** [comment]
Hey @onyxcherry sorry for the huge delay!

> Although the message could be improved by implementing own Deserialize for ArrayOrSingleObject, isn't it better to simply replace the message: [...]

I don't like to match on a string that could change at any time if serde decides to update their errors. But I like your message I think we could use it in any case. Do you think we want to let the original serde message sometimes?
Or maybe we could look for the `untagged` word in the error message and only override it in this case. Because that's the main point causing the issue.

**Comment by onyxcherry** [comment]
0. `untagged` attribute which - as you say - causes the issue appears only once in the meilisearch code so perhaps the error message provided by serde should be (completely) changed only there,
1. User data are validated against untagged enum in the _whichever matches first, we return_ (from a parsing function) way,
2. So the best, explanatory and helpful error messages are these listing all the options of the Enum which could be matched and parsed correctly.

Is [serde_json::error::Category](https://docs.rs/serde_json/latest/serde_json/error/enum.Category.html#) (which can be obtained with `.classify()` on `serde_json::Error`) the optimal way to decide if the error message should be replaced by us?

See its usage on common examples:

```rust
use either::Either;
use serde::Deserialize;
use serde_json;

type Object = serde_json::Map<String, serde_json::Value>;

#[derive(Deserialize, Debug)]
#[serde(transparent)]
#[allow(dead_code)]
struct ArrayOrSingleObject {
    #[serde(with = "either::serde_untagged")]
    inner: Either<Vec<Object>, Object>,
}

fn main() {
    let payload_syntax_err = "[{ \"id\": 1, \"pet\" \"dog\" }]"; // <-- note missing semicolon
    let result1: Result<ArrayOrSingleObject, serde_json::Error> = serde_json::from_str(payload_syntax_err);
    let error1 = result1.err().unwrap();
    println!("{:#?}", error1.classify()); // Syntax

    let payload_data_err = "\"some string\"";
    let result2: Result<ArrayOrSingleObject, serde_json::Error> = serde_json::from_str(payload_data_err);
    let error2 = result2.err().unwrap();
    println!("{:#?}", error2.classify()); // Data
}
```

Note that if the `Category` is `Syntax`, [`serde_msg`](https://github.com/meilisearch/meilisearch/blob/2827ff7957675c3deb261d582168d1ca4cf6fe11/meilisearch-lib/src/document_formats.rs#L45) contains helpful _Couldn't serialize document value: expected `:` at line 1 column 19_ message.

To sum up, my proposal is the change of the meaningful message _only_ if the `Category` of the serde error is `Data`. \
And in that case the best response we can return is a list of expected objects or the error saying _neither this nor that_ (e.g. my previous one ^).

**Comment by irevoire** [comment]
Hoy @onyxcherry!
> Is [serde_json::error::Category](https://docs.rs/serde_json/latest/serde_json/error/enum.Category.html#) (which can be obtained with .classify() on serde_json::Error) the optimal way to decide if the error message should be replaced by us?

I didn't know this existed; it's perfect for us. Thanks for finding that out!
Do you want to implement it?
